### PR TITLE
Fix unnamed lifetime spans

### PR DESCRIPTION
**Pull request by bovinebuddha**
Fixes issue #125143.

The error message used to mix up two different unnamed lifetimes since they were both assigned the name "'_". Looking up the span using the DefId instead gave correct spans.

I am not sure if a similar fix should be made for any of the other RegionKinds, e.g. ReLateParam, and I am unsure of how to construct a test case which used late bound lifetimes or RePlaceholder.

Maybe the function 'msg_span_from_named_region' should be renamed since it now uses (partially) DefId instead of the name. Also, the name is quite misleading since it is sometimes used to look up unnamed lifetimes...

### CODE REVIEW

File: compiler/rustc_hir/src/hir.rs [hunk 1]
```diff
@@ -571,6 +571,10 @@ impl<'hir> Generics<'hir> {
         self.params.iter().find(|&param| name == param.name.ident().name)
     }
 
+    pub fn get_with_def_id(&self, id: DefId) -> Option<&GenericParam<'hir>> {
```
Comment (compiler-errors) [inline compiler/rustc_hir/src/hir.rs:574]: This is definitely not correct if `DefId` comes from a different crate. Please make this take a `LocalDefId`, or ideally, do the comparison with `HirId` instead.

**Review by compiler-errors — CHANGES_REQUESTED** [review]
See comment

File: compiler/rustc_hir/src/hir.rs (hunk 1 above)
Comment (bovinebuddha) [reply compiler/rustc_hir/src/hir.rs]: I wad not able to find a way to convert a DefId to a HirId, so I used a LocalDefId instead. The place where I use it already has expect_local() on the binding scope, so I assume that the lifetime must also be local.

**Comment by alex-semenyuk** [comment]
@bovinebuddha 
Form wg-triage. Closed this PR due to inactivity. Feel free to reopen or raised new one. Thanks for your efforts.

**Pull request closed without merging**